File: java/src/main/java/cn/az/code/year2020/nov/InsertionSortList.java
package cn.az.code.year2020.nov;

import cn.az.code.common.ListNode;

/**
 * @author az
 * @since 11/02/20
 */
public class InsertionSortList {

    public ListNode sort(ListNode head) {

        if (head == null || head.next == null) {
            return head;
        }
        ListNode dummy = new ListNode(0);
        ListNode prev = dummy;

        while (head != null) {
            ListNode temp = head.next;

            if (prev.val >= head.val) {
                prev = dummy;
            }
            while (prev.next != null && prev.next.val < head.val) {
                prev = prev.next;
            }

            head.next = prev.next;
            prev.next = head;
            head = temp;
        }

        return dummy.next;
    }
}

File: archive/cpp/2021/dec/insertion-sort-list.cc
#include "common.h"

class Solution
{
public:
    ListNode *insertionSortList(ListNode *head)
    {
        ListNode *dummy = new ListNode(0);
        dummy->next = head;

        ListNode *pre = dummy, *cur = head;

        while (cur)
        {
            if ((cur->next) && (cur->next->val < cur->val))
            {
                while ((pre->next) && (pre->next->val < cur->next->val))
                {
                    pre = pre->next;
                }
                ListNode *tmp = pre->next;
                pre->next = cur->next;
                cur->next = cur->next->next;
                pre->next->next = tmp;
                pre = dummy;
            }
            else
            {
                cur = cur->next;
            }
        }
        return dummy->next;
    }
};